zrml/global-disputes/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+// your option) any later version.
+//
+// Zeitgeist is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
+#![doc = include_str!("../README.md")]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod benchmarks;
+mod global_disputes_pallet_api;
+mod mock;
+mod tests;
+pub mod weights;
+
+pub use global_disputes_pallet_api::GlobalDisputesPalletApi;
+pub use pallet::*;
+
+#[frame_support::pallet]
+mod pallet {
+    use crate::{weights::WeightInfoZeitgeist, GlobalDisputesPalletApi};
+    use core::marker::PhantomData;
+    use frame_support::{
+        ensure, log,
+        pallet_prelude::{
+            Decode, DispatchError, DispatchResultWithPostInfo, Encode, MaxEncodedLen, OptionQuery,
+            StorageDoubleMap, StorageMap, TypeInfo, ValueQuery,
+        },
+        sp_runtime::{traits::StaticLookup, RuntimeDebug},
+        storage::child::KillStorageResult,
+        traits::{
+            Currency, ExistenceRequirement, Get, IsType, LockIdentifier, LockableCurrency,
+            WithdrawReasons,
+        },
+        Blake2_128Concat, BoundedVec, PalletId, Twox64Concat,
+    };
+    use frame_system::{ensure_signed, pallet_prelude::OriginFor};
+    use sp_runtime::traits::{AccountIdConversion, CheckedDiv, Saturating, Zero};
+    use sp_std::{vec, vec::Vec};
+    use zeitgeist_primitives::types::OutcomeReport;
+    use zrml_market_commons::MarketCommonsPalletApi;
+
+    pub(crate) type BalanceOf<T> = <<T as Config>::Currency as Currency<AccountIdOf<T>>>::Balance;
+    pub(crate) type MarketIdOf<T> =
+        <<T as Config>::MarketCommons as MarketCommonsPalletApi>::MarketId;
+    pub type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+    pub type OutcomeInfoOf<T> = OutcomeInfo<BalanceOf<T>, OwnerInfoOf<T>>;
+    pub type WinnerInfoOf<T> = WinnerInfo<BalanceOf<T>, OwnerInfoOf<T>>;
+    type AccountIdLookupOf<T> = <<T as frame_system::Config>::Lookup as StaticLookup>::Source;
+    type OwnerInfoOf<T> = BoundedVec<AccountIdOf<T>, <T as Config>::MaxOwners>;
+    type LockInfoOf<T> = LockInfo<MarketIdOf<T>, BalanceOf<T>>;
+
+    #[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+    pub struct LockInfo<MarketId, Balance>(pub Vec<(MarketId, Balance)>);
+
+    impl<MarketId, Balance> Default for LockInfo<MarketId, Balance> {
+        fn default() -> Self {
+            Self(Vec::new())
+        }
+    }
+
+    impl<MarketId, Balance> MaxEncodedLen for LockInfo<MarketId, Balance>
+    where
+        Balance: MaxEncodedLen,
+        MarketId: MaxEncodedLen,
+    {
+        fn max_encoded_len() -> usize {
+            MarketId::max_encoded_len().saturating_add(Balance::max_encoded_len())
```

Comment:
Thank you for explaining `max_encoded_len`. This will be deprecated, because I will use a `BoundedVec` instead. I keep this conversation unresolved to not forget it.

Reply:
e00f66217d5da65308d3b3096517897d0441e917